File: moringacodilityquestions.js
// Write a function that merges two strings together. It does so by merging the end of the first string with the start of the second string together when they are an exact match.

// "abcde" + "cdefgh" => "abcdefgh"

// "abaab" + "aabab" => "abaabab"

// "abc" + "def" => "abcdef"

// "abc" + "abc" => "abc"

// NOTE: The algorithm should always use the longest possible overlap.

// "abaabaab" + "aabaabab" would be "abaabaabab" and not "abaabaabaabab"


// 1. 

// In the context of the given problem, an overlap refers to the number of characters at the end of the first string 
// that match the beginning of the second string.

// For example, given the strings "abcde" and "cdefgh", the overlap is "cde" (the 3 characters at the end of "abcde" 
// that match the 3 characters at the start of "cdefgh").

// The merge_strings() function uses the overlap to determine how much of the second string should be appended to the first 
// string in order to create the merged string. It ensures that the overlapping part is only included once in the merged string,
//  and the algorithm always tries to use the longest possible overlap.

// 2.
// suppose you have two strings s1 and s2. You want to merge them into a single string such that the end of 
// s1 matches the start of s2.

// Here are some examples to illustrate:

// "abcde" and "cdefgh" can be merged into "abcdefgh". Notice how "cde" is common to both strings, 
// so we only include it once in the merged string.

// "abaab" and "aabab" can be merged into "abaabab". Notice how "aab" is common to both strings, 
// but "aba" is also common. However, we need to use the longest possible overlap, which is "aab", 
// to avoid duplicating characters in the merged string.

// "abc" and "def" can be merged into "abcdef". There is no overlap between these two strings, so we just concatenate them.

// "abc" and "abc" can be merged into "abc". The two strings are identical, so we just return one of them.

// In general, the function should use the longest possible overlap between s1 and s2 to create the merged string, 
// to avoid duplicating any characters in the merged string.


function mergeStrings(s1, s2) {
    let overlap = 0;
    for (let i = 1; i <= Math.min(s1.length, s2.length); i++) {
      if (s1.slice(-i) === s2.slice(0, i)) {
        overlap = i;
      }
    }
    return s1 + s2.slice(overlap);
  }

  
  console.log(mergeStrings("abaabaab", "aabaabab"));  // output: "abaabaabab"

File: moringacodilityquestions.rb
# 1. function takes two numbers => def compareNumbers(a,b)
# 2. when a > b -> a is greater than b 
# 3. when a < b -> a is less than b 
# 4. else a = b -> a is equal than b 

# def compareNumbers(a,b)
#     #using string interpolation
# if 'a' is equal 'b' return a string "equal to"
#     ....
# end

# <=> : spaceship conditional.

def compareNumbers(a,b)
    comparison = a <=> b
    abs_comparison = comparison.abs 
    str = "#{a} is "
    str += "greater than " * ((comparison + 1) / 2)
    str += "less than " * ((-comparison + 1) / 2)
    str += "equal to " * (1 - abs_comparison)
    str += "#{b}"
 end
 
 puts compareNumbers(2,2)
 puts compareNumbers(4,2)
 puts compareNumbers(2,4)
 
 
 # 2. 
 # BDD 
 # countCharacters('Bass')
 # [["B",1],["a",1],["s",2]]
 # 1. def countCharacters('')
 #     have an empty hash => keep count of each character in the input string -< {}
 #     have an empty array => this array will contain hashes with key ->character , count -> value 
 #     loop through each character in the string , -> each -> each_char(loop,sort)
 #     if the hash does not contain the character as a key , we create it and set its count to 0 
 #     each subsequent iteration will increment the value of the key 
 #     loop the hash , append an array of the key and value to our empty array 
 #     return a list of arrays 
 # end 
 
 def count_characters(str)
     char_count = {} 
     result = []
     str.each_char do |c|
         char_count[c] ||= 0
         char_count[c] += 1
     end
     char_count.each_pair do |key,value|
         result << [key,value]
     end
     result 
 end
 
 puts count_characters("Bass")


#  2. // Write a function that merges two strings together. It does so by merging the end of the first string with the start of the second string together when they are an exact match.

#  // "abcde" + "cdefgh" => "abcdefgh"
 
#  // "abaab" + "aabab" => "abaabab"
 
#  // "abc" + "def" => "abcdef"
 
#  // "abc" + "abc" => "abc"
 
#  // NOTE: The algorithm should always use the longest possible overlap.
 
#  // "abaabaab" + "aabaabab" would be "abaabaabab" and not "abaabaabaabab"

# check explanation for pseudocode on .js file

def merge_strings(s1, s2)
    overlap = 0
    (1..[s1.length, s2.length].min).each do |i|
      if s1[-i..-1] == s2[0...i]
        overlap = i
      end
    end
    s1 + s2[overlap..-1]
  end
  

  puts merge_strings("abc", "abc")  # output: "abcdefgh"
# syntax explanation: 

# In the expression s1[-i..-1], the .. operator creates a range object that includes the last i characters of s1.
#  The -1 argument means that the end of the range is the last character of s1. The negative index is used to count the positions
#   from the end of the string, so s1[-i] would be the i-th character from the end.

# Similarly, in the expression s2[0...i], the ... operator creates a range object that includes the first i characters of s2. 
# The 0 argument means that the start of the range is the first character of s2. The ... operator is used to exclude the i-th character from the range.

# So, the expression s1[-i..-1] gets the last i characters of s1, and the expression s2[0...i] gets the first i characters of s2. 
# The if statement then checks if these two substrings are equal, which means they are a possible overlap.

# camel case question 
# Challenge 1: CamelCase Method(Toy Problem)
# Write simple .camelCase method for strings. All words must have their first letter capitalized without spaces.

# For instance:

# 'hello case'.camelcase => HelloCase

# 'camel case word'.camelcase => CamelCaseWord
# 1. first split the string into an array of words using the split method
# 2. map over each word to capitalize the first letter using capitalize
# 3. join the words back into a single string using the join method.
def camelCase(a)
  a.split.map(&:capitalize).join(' ')
end 

puts camelCase("joseph mbugua")
